layout: Rewrite Servo's `vertical-align` support to match CSS 2.1 § 10.8, and implement `vertical-align: middle` per CSS 2.1 § 10.8.1.

pcwalton opened the pull request:
`InlineMetrics` has been split into `InlineMetrics` for fragments and
`LineMetrics` for lines. Both structures' fields have been renamed in
order to more clearly delineate the difference between _space_ and
_content_. Vertical positioning of fragments has been reworked to take
margins and borders into account only for replaced content.

This patch fixes the `vertical_align_super_a.html` reftest. Servo now
matches the rendering that Gecko and WebKit produce.

Additionally, this includes a test for the popular inline-block
centering technique described here:
https://s.codepen.io/shshaw/fullpage/gEiDt?#Inline-Block

r? @mbrubeck

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13602)

<!-- Reviewable:end -->

pcwalton committed:
layout: Rewrite Servo's `vertical-align` support to match CSS 2.1 §
10.8, and implement `vertical-align: middle` per CSS 2.1 § 10.8.1.

`InlineMetrics` has been split into `InlineMetrics` for fragments and
`LineMetrics` for lines. Both structures' fields have been renamed in
order to more clearly delineate the difference between *space* and
*content*. Vertical positioning of fragments has been reworked to take
margins and borders into account only for replaced content.

This patch fixes the `vertical_align_super_a.html` reftest. Servo now
matches the rendering that Gecko and WebKit produce.

Additionally, this includes a test for the popular inline-block
centering technique described here:
https://s.codepen.io/shshaw/fullpage/gEiDt?#Inline-Block

pcwalton commented:
r? @mbrubeck

emilio commented:
@bors-servo: try

I'm surprised not to see the `x_height` of the font involved here, since it's explicitly mentioned in the spec, but maybe you're doing something smart with the font metrics?

bors-servo commented:
:hourglass: Trying commit faef393 with merge 336f014...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/48)

highfive commented:
▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox-flex-direction-default.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox-flex-direction-row.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox-flex-direction-row-reverse.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox-flex-wrap-wrap-reverse.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox-flex-wrap-wrap.htm

pcwalton commented:
Updated to mark tests as newly passing. r? @mbrubeck

mbrubeck commented on components/layout/inline.rs:
```diff
@@ -1069,7 +1069,21 @@ impl InlineFlow {
+                        // (That is, subtract out the pre-existing `block_start` value.)
+                        block_start = block_start -
+                            line_block_metrics.size_above_baseline +
+                            line_block_metrics.size.scale_by(0.5) +
```
As @emilio said, I think this should use the x-height.

mbrubeck commented on components/layout/inline.rs:
```diff
@@ -1069,7 +1069,21 @@ impl InlineFlow {
+                            line_block_metrics.size_above_baseline +
+                            line_block_metrics.size.scale_by(0.5) +
+                            fragment_inline_metrics.ascent.scale_by(0.5) -
+                            fragment_inline_metrics.depth_below_baseline.scale_by(0.5)
```
Should this be `fragment.border_box.size.block.scale_by(0.5)` to find the middle of the fragment?  Currently if the fragment is an `<img>` whose height is different from the line height, it doesn't align correctly.

pcwalton committed:
resources: Remove the `vertical-align: middle` rule for buttons from
Servo's style sheet.

This doesn't vertically center the contents of buttons as it's intended
to and breaks layouts once `vertical-align: middle` is implemented.

pcwalton committed:
layout: Take the position of `InlineFlow`s in account when determining
the baseline offset of the last line box in flow.

This caused incorrect layouts in some cases.

pcwalton committed:
gfx: On the Mac, take the scale into account when determining the
x-height of a font.

pcwalton commented:
Here's the rewrite! See commits for details.

r? @mbrubeck

frewsxcv commented:
```
The command "bash etc/ci/manifest_changed.sh" exited with 1.
```

jdm commented:
Looks like a bunch of missing reftest `<link>` attributes.

mbrubeck commented:
Looks good! r=mbrubeck with one minor improvement suggestion (below).

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 13 files at r1, 12 of 12 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 35 of 35 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/layout/fragment.rs, line 2244 at r5](https://reviewable.io:443/reviews/servo/servo/13602#-KToplC6EG1AecPt_eyO:-KToplC7DtyDrxYRh8OV:b-n556dh) ([raw file](https://github.com/servo/servo/blob/15ed039856bb7c7df97f8217ddece64b8e046ba2/components/layout/fragment.rs#L2244)):*
> ```Rust
>                                  actual_line_metrics: Option<&LineMetrics>)
>                                  -> Au {
>         // FIXME(pcwalton): Cache `self.content_inline_metrics()` if we need it multiple times?
> ```

You could just pass in the `content_inline_metrics` from the caller (`aligned_inline_metrics`, below).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13602#-:-KToqpPc5ZtlxYQ_nN9B:btbk07k)*
<!-- Sent from Reviewable.io -->

pcwalton commented:
@bors-servo: r=mbrubeck

bors-servo commented:
:pushpin: Commit 773614f has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 773614f84c679aac2a7fd455f706e59d8222a967 -->

bors-servo commented:
:hourglass: Testing commit 773614f with merge 768a699...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3462), [arm64](http://build.servo.org/builders/arm64/builds/3474), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3547), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/168), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/169), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3539), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3518), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/220), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/220), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1427)
